### PR TITLE
cherry-pick(v12.1.1):  Bump `@metamask/eth-json-rpc-middleware` to `^14.0.0` (#26143)

### PR DESCRIPTION
**Pull request by MajorLift**
## **Description**

Cherry-pick of #26143 for ~v12.1.0-rc~ v12.1.1-rc.

This is a very different PR from the original, as I had to remove diffs to a significant amount of code that has not yet been introduced to the release candidate.

The diffs that were excluded here will need to be reapplied to a future release. I'm also keeping an eye out for new cherry-picks that may introduce changes requiring me to restore certain diffs.

Aligning `@metamask/eth-block-tracker` to `^11.0.1` is deferred, as it's blocked by #26150,  which is not included in this release candidate. The `@metamask/transaction-controller` major version bump from `^34.0.0` to `^35.1.1` is also removed for now. 

See below for new changelog.

### Added

- Add and export `PPOMMiddlewareRequest` type for `JsonRpcRequest` types
that include the `securityAlertResponse` property.
  - `securityAlertResponse` is defined as both optional and nullable.
- Add `PPOMRequest` type for `eth-sendTransaction` requests.

### Changed

- **BREAKING:** Bump `@metamask/eth-json-rpc-middleware` from `^12.1.1`
to `^14.0.0`.
- Bump `@trezor/connect-web` from `9.2.2` to `9.3.0`.

### Fixed

- **BREAKING:** Narrow `Params` generic parameter of
`createPPOMMiddleware` function from `JsonRpcParams` to `(string | { to:
string })[]`.
- Add `Params` generic parameter to `handleSnapRequest` function, which
is constrained by `Record<string, unknown>` and defaults to `JsonRpcParams`.
  - `handleSnapRequest` can now be typed correctly with any `params`
object.

### Security

- **BREAKING:** Typed signature validation only replaces `0X` prefix
with `0x`, and contract address normalization is removed for decimal and
octal values.
  - Threat actors have been manipulating `eth_signTypedData_v4` fields to
cause failures in blockaid's detectors.
  - Extension crashes with an error when performing Malicious permit with
a non-0x prefixed integer address.
  - This fixes an issue where the key value row or petname component
disappears if a signed address is prefixed by "0X" instead of "0x".

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@SocketSecurity ignore npm/@storybook/node-logger@7.6.20 npm/@storybook/client-logger@7.6.20 npm/@storybook/postinstall@7.6.20 npm/@storybook/react-dom-shim@7.6.20 npm/@storybook/react@7.6.20 npm/@storybook/addon-actions@7.6.20 npm/@storybook/addon-docs@7.6.20 npm/@storybook/core-common@7.6.20 npm/@storybook/manager-api@7.6.20 npm/@storybook/preview-api@7.6.20 npm/@storybook/core-events@7.6.20 npm/@storybook/blocks@7.6.20 npm/@storybook/components@7.6.20 npm/@storybook/csf-plugin@7.6.20 npm/@storybook/csf-tools@7.6.20 npm/@storybook/theming@7.6.20 npm/@storybook/types@7.6.20 npm/@storybook/channels@7.6.20 npm/@storybook/docs-tools@7.6.20 npm/@storybook/router@7.6.20 npm/@storybook/core-client@7.6.20

Patch updates auto-applied due to caret specifier in version string: `^7.6.19`.

**Comment by MajorLift** [comment]
@SocketSecurity ignore npm/diff@5.2.0

Benign author change

**Comment by MajorLift** [comment]
@SocketSecurity ignore npm/axios@1.7.4

Library functionality requires network access.

**Comment by MajorLift** [comment]
@metamask update-policies

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by Gudahtt** [comment]
We'll be targeting this for v12.1.1 instead, so I've temporarily labelled this as `DO-NOT-MERGE` until that RC is ready to target.

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@SocketSecurity ignore npm/@storybook/addon-actions@7.6.20, npm/@storybook/addon-docs@7.6.20, npm/@storybook/blocks@7.6.20, npm/@storybook/channels@7.6.20, npm/@storybook/client-logger@7.6.20, npm/@storybook/components@7.6.20, npm/@storybook/core-client@7.6.20, npm/@storybook/core-common@7.6.20, npm/@storybook/core-events@7.6.20, npm/@storybook/csf-plugin@7.6.20, npm/@storybook/csf-tools@7.6.20, npm/@storybook/docs-tools@7.6.20, npm/@storybook/manager-api@7.6.20, npm/@storybook/node-logger@7.6.20, npm/@storybook/postinstall@7.6.20, npm/@storybook/preview-api@7.6.20, npm/@storybook/react-dom-shim@7.6.20, npm/@storybook/react@7.6.20, npm/@storybook/router@7.6.20, npm/@storybook/theming@7.6.20, npm/@storybook/types@7.6.20

Patch version bump for trusted `@storybook`-namespaced packages.

**Comment by MajorLift** [comment]
@SocketSecurity ignore npm/@storybook/addon-actions@7.6.20, npm/@storybook/addon-docs@7.6.20, npm/@storybook/blocks@7.6.20, npm/@storybook/channels@7.6.20, npm/@storybook/client-logger@7.6.20, npm/@storybook/components@7.6.20, npm/@storybook/core-client@7.6.20, npm/@storybook/core-common@7.6.20, npm/@storybook/core-events@7.6.20, npm/@storybook/csf-plugin@7.6.20, npm/@storybook/csf-tools@7.6.20, npm/@storybook/docs-tools@7.6.20, npm/@storybook/manager-api@7.6.20, npm/@storybook/node-logger@7.6.20, npm/@storybook/postinstall@7.6.20, npm/@storybook/preview-api@7.6.20, npm/@storybook/react-dom-shim@7.6.20, npm/@storybook/react@7.6.20, npm/@storybook/router@7.6.20, npm/@storybook/theming@7.6.20, npm/@storybook/types@7.6.20

Patch version bump for trusted `@storybook`-namespaced packages.

**Comment by MajorLift** [comment]
https://github.com/SocketSecurity ignore npm/diff@5.2.0

Benign author change

**Comment by MajorLift** [comment]
https://github.com/SocketSecurity ignore npm/axios@1.7.4

Library functionality requires network access.

**Comment by MajorLift** [comment]
@SocketSecurity ignore npm/@storybook/addon-toolbars@7.6.20 npm/@storybook/preview@7.6.20 npm/@storybook/manager@7.6.20 npm/@storybook/addon-essentials@7.6.20 npm/storybook@7.6.20 npm/@storybook/addon-backgrounds@7.6.20 npm/@storybook/addon-controls@7.6.20 npm/@storybook/addon-highlight@7.6.20 npm/@storybook/addon-measure@7.6.20 npm/@storybook/addon-outline@7.6.20 npm/@storybook/addon-viewport@7.6.20 npm/@storybook/builder-webpack5@7.6.20 npm/@storybook/preset-react-webpack@7.6.20 npm/@storybook/core-webpack@7.6.20 npm/@storybook/cli@7.6.20 npm/@storybook/codemod@7.6.20 npm/@storybook/core-server@7.6.20 npm/@storybook/telemetry@7.6.20 npm/@storybook/addon-a11y@7.6.20 npm/@storybook/builder-manager@7.6.20 npm/@storybook/react-webpack5@7.6.20 npm/@storybook/addons@7.6.20 npm/@storybook/api@7.6.20 npm/@storybook/addon-mdx-gfm@7.6.20 npm/@storybook/client-api@7.6.20

Patch version bump for trusted @storybook-namespaced packages.

File: yarn.lock [hunk 1]
```diff
@@ -8213,7 +8143,7 @@ __metadata:
   languageName: node
   linkType: hard
 
-"@storybook/addon-actions@npm:7.6.19, @storybook/addon-actions@npm:^7.6.19":
+"@storybook/addon-actions@npm:7.6.19":
```
[the remark this answers is not in the document]
Comment (MajorLift) [reply yarn.lock:8146]: Good point! Fixed here https://github.com/MetaMask/metamask-extension/pull/26626/commits/0d1f3c842dce148134a3f24de4652bdde45cb3ed

File: test/jest/mocks.js [hunk 2]
```diff
@@ -172,10 +172,10 @@ export const getInitialSendStateWithExistingTxState = (draftTxState) => ({
 
 export function createMockInternalAccount({
   address = MOCK_DEFAULT_ADDRESS,
-  name,
+  name = MOCK_NAME,
   type = EthAccountType.Eoa,
   keyringType = KeyringTypes.hd,
-  snapOptions,
+  snapOptions = { id: '', name: '', enabled: true },
```
Comment (MajorLift) [inline test/jest/mocks.js:178]: Are these acceptable defaults?

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Review by Mrtenz — DISMISSED** [review]

File: app/scripts/migrations/105.ts [hunk 3]
```diff
@@ -103,6 +103,7 @@ function createInternalAccountsForAccountsController(
           // initial updateAccounts call.
           type: 'HD Key Tree',
         },
+        importTime: Date.now(),
```
Comment (Gudahtt) [inline app/scripts/migrations/105.ts:106]: Looks like this migration change, and the 119 migration change, were caused by a dependency bump that we can remove from the lock file

File: .yarn/patches/@trezor-connect-web-npm-9.3.0-040ab10d9a.patch
Comment (Gudahtt) [inline .yarn/patches/@trezor-connect-web-npm-9.3.0-040ab10d9a.patch:1]: We forgot to delete the old patch file, but that's OK for the cherry-pick, it'll just get ignored. We can follow up to delete it on `develop` at least though.

File: app/scripts/lib/createDupeReqFilterStream.ts [hunk 4]
```diff
@@ -54,7 +54,8 @@ export default function createDupeReqFilterStream() {
     transform(chunk: JsonRpcRequest, _, cb) {
       // JSON-RPC notifications have no ids; our only recourse is to let them through.
       const hasNoId = chunk.id === undefined;
-      const requestNotYetSeen = seenRequestIds.add(chunk.id);
+      const requestNotYetSeen =
+        chunk.id !== null && seenRequestIds.add(chunk.id);
```
Comment (Gudahtt) [inline app/scripts/lib/createDupeReqFilterStream.ts:58]: I'm a bit concerned that this might have a functional impact, for requests with an id of `null`. I don't know why we'd get requests like that, but, I wasn't able to rule it out. I think we can allow null as a seen request ID, rather than adding a condition for it here.

**Review by Gudahtt — COMMENTED** [review]
Just had a call where we discussed a few changes, I left some comments so there'd be some context for anyone looking back on this.

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@metamaskbot update-policies

File: .yarn/patches/@trezor-connect-web-npm-9.3.0-040ab10d9a.patch
Comment (MajorLift) [reply .yarn/patches/@trezor-connect-web-npm-9.3.0-040ab10d9a.patch:1]: - Created a PR for removing the old patch file in `develop`: https://github.com/MetaMask/metamask-extension/pull/26814

File: app/scripts/lib/createDupeReqFilterStream.ts (hunk 4 above)
Comment (MajorLift) [reply app/scripts/lib/createDupeReqFilterStream.ts]: - Reverted this change: fe7c6ad17978cf7d8f60d25e46f6ccffe1b1f7d4
- Created a PR for reverting this in `develop`: https://github.com/MetaMask/metamask-extension/pull/26814

File: test/jest/mocks.ts [hunk 5]
```diff
@@ -171,19 +176,28 @@ export const getInitialSendStateWithExistingTxState = (draftTxState) => ({
 });
 
 export function createMockInternalAccount({
```
Comment (MajorLift) [inline test/jest/mocks.ts:178]: I had to partially cherry-pick commit in `develop` that converts `test/jest/mocks.js` into TypeScript, because the params types of `createMockInternalAccount` was being incompletely inferred, causing conflicting type errors downstream. Supplying the correct type was the only solution.

File: yarn.lock
Comment (Gudahtt) [inline yarn.lock:1]: Hmm, I'm seeing a lot of changes here that seem unrelated to the changes. When I run `git checkout origin/Version-v12.1.1 yarn.lock && yarn && yarn dedupe` it seems to eliminate most of them.

Comment (Gudahtt) [reply yarn.lock:1]: Looks like no type errors are introduced by this either, except one `ts-expect-error` comment that can be removed (in `app/scripts/lib/ppom/ppom-middleware.test.ts`)

**Review by Gudahtt — DISMISSED** [review]
Mostly looks good! But I think we can trim down these lock file changes further

**Comment by MajorLift** [comment]
@metamaskbot update-policies

File: yarn.lock
Comment (MajorLift) [reply yarn.lock:1]: That's a huge difference! I'll have to remember to check for unnecessary lockfile diffs. 

- lockfile trimmed: https://github.com/MetaMask/metamask-extension/pull/26626/commits/6d45f37fb9e504f0068d8c2e3fd8dcd5473c9b46
- unused `ts-expect-error` removed: https://github.com/MetaMask/metamask-extension/pull/26626/commits/6a41dc40e04c9ab84a3adf3490a271862ef81352

**Review by Gudahtt — APPROVED** [review]
LGTM!
